go/libraries/doltcore/sqle/binlogreplication/binlog_primary_streamer.go

```diff
@@ -137,20 +234,108 @@ func (m *binlogStreamerManager) copyStreamers() []*binlogStreamer {
 	streamer := newBinlogStreamer()
 	m.addStreamer(streamer)
 	defer m.removeStreamer(streamer)
 
-	return streamer.startStream(ctx, conn, binlogFormat, binlogEventMeta)
+	file, err := m.findLogFileForPosition(executedGtids)
+	if err != nil {
+		return err
+	}
+
+	return streamer.startStream(ctx, conn, executedGtids, binlogFormat, &binlogEventMeta, file)
 }
 
-// sendEvents sends |binlogEvents| to all the streams managed by this instance.
-func (m *binlogStreamerManager) sendEvents(binlogEvents []mysql.BinlogEvent) {
-	for _, streamer := range m.copyStreamers() {
-		logrus.StandardLogger().Tracef("queuing %d binlog events\n", len(binlogEvents))
-		streamer.eventChan <- binlogEvents
+// findLogFileForPosition searches through the available binlog files on disk for the first log file that
+// contains GTIDs that are NOT present in |executedGtids|. This is determined by reading the first GTID event
+// from each log file and selecting the previous file when the first GTID not in |executedGtids| is found. If
+// the first GTID event in all available logs files is in |executedGtids|, then the current log file is returned.
+func (m *binlogStreamerManager) findLogFileForPosition(executedGtids mysql.GTIDSet) (string, error) {
+	files, err := m.logManager.logFilesOnDiskForBranch(BinlogBranch)
+	if err != nil {
+		return "", err
+	}
+
+	for i, f := range files {
+		binlogFilePath := filepath.Join(m.logManager.binlogDirectory, f)
+		file, err := openBinlogFileForReading(binlogFilePath)
+		if err != nil {
+			return "", err
+		}
+
+		binlogEvent, err := readFirstGtidEventFromFile(file)
+		if fileCloseErr := file.Close(); fileCloseErr != nil {
+			logrus.Errorf("unable to cleanly close binlog file %s: %s", f, err.Error())
+		}
+		if err == io.EOF {
+			continue
+		} else if err != nil {
+			return "", err
+		}
+
+		if binlogEvent.IsGTID() {
+			gtid, _, err := binlogEvent.GTID(m.logManager.binlogFormat)
+			if err != nil {
+				return "", err
+			}
+			// If the first GTID in this file is contained in the set of GTIDs that the replica
+			// has already executed, then move on to check the next file.
+			if executedGtids.ContainsGTID(gtid) {
+				continue
+			}
+
+			// If we found an unexecuted GTID in the first binlog file, return the first file,
+			// otherwise return the previous file
+			if i == 0 {
+				return binlogFilePath, nil
+			} else {
+				return filepath.Join(m.logManager.binlogDirectory, files[i-1]), nil
+			}
+		}
+	}
+
+	// If we don't find any GTIDs that are missing from |executedGtids|, then return the current
+	// log file so the streamer can reply events from it, potentially finding GTIDs later in the
+	// file that need to be sent to the connected replica, or waiting for new events to be written.
+	return m.logManager.currentBinlogFilepath(), nil
+}
+
+func readBinlogEventFromFile(file *os.File) (mysql.BinlogEvent, error) {
```

Comment:
Seems like this logic would be better off in a BinLogEventReader, rather than the streamer